web/app/services/sitemap.service.js

```diff
@@ -0,0 +1,182 @@
+            if (!cols) cols = 12;
+            if (!row) row = 0;
+            if (!col) col = 0;
+
+            angular.forEach(srcwidgets, function (widget) {
+                var w;
+                switch (widget.type) {
+                    case "Frame":
+                        w = {
+                            id: widget.widgetId,
+                            sizeX: 12,
+                            sizeY: 1,
+                            type: 'label',
+                            name: widget.label,
+                            background: 'black'
+                        };
+                        break;
+                    case "Colorpicker":
+                        w = {
+                            id: widget.widgetId,
+                            sizeX: 4,
+                            sizeY: 4,
+                            type: 'colorpicker',
+                            name: widget.label,
+                            item: (widget.item) ? widget.item.name : null
+                        };
+                        break;
+                    case "Slider":
+                        w = {
+                            id: widget.widgetId,
+                            sizeX: 4,
+                            sizeY: 4,
+                            type: 'slider',
+                            name: widget.label,
+                            item: (widget.item) ? widget.item.name : null
+                        };
+                        break;
+                    case "Setpoint":
+                        w = {
+                            id: widget.widgetId,
+                            sizeX: 4,
+                            sizeY: 4,
+                            type: 'knob',
+                            name: widget.label,
+                            item: (widget.item) ? widget.item.name : null
+                        };
+                        break;
+                    case "Switch":
+                        w = {
+                            id: widget.widgetId,
+                            sizeX: 4,
+                            sizeY: 4,
+                            type: 'switch',
+                            name: widget.label,
+                            item: (widget.item) ? widget.item.name : null,
+                            iconset: (widget.icon) ? 'eclipse-smarthome-classic' : undefined,
+                            icon: (widget.icon) ? widget.icon : undefined,
+                        };
+                        break;
+                    case "Image":
+                        w = {
+                            id: widget.widgetId,
+                            sizeX: 4,
+                            sizeY: 4,
+                            type: 'image',
+                            name: widget.label,
+                            url: widget.url
+                        };
+                        break;
+                    case "Webview":
+                        w = {
```

Comment:
Maybe instead of duplicating the same properties on each case, you could define only specific key-values there:
```javascript
    type: 'frame',
    frameUrl: widget.url
```

And then after the `switch` statement you could extend the object. You could use lodash:
```
    _.extend(w, {
        id: widget.widgetId,
        sizeX: 4,
        sizeY: 4,
        name: widget.label
    });
```

I was thinking about including lodash to our `vendor.js` anyway.
Lodash exists already in bower_components (or node_modules, don't remember atm), so we could include it inside `gulpfile.js` `vendor-js` task.